scalyr_agent/copying_manager/worker.py

```diff
@@ -288,6 +288,16 @@ def log_worker_session_status(self):
         """
         pass
 
+    @abstractmethod
+    def get_full_checkpoints(self):  # type: () -> Dict
+        """
+        This method returns the full collection of the current log processor checkpoint states.
```

Comment:
There's a subtle distinction that I think we need to dig into.  Should this return only the checkpoints for what the worker is currently processing, or all checkpoints from its checkpoint file?  From my understanding, the checkpoints here will be used when opening a file to start processing it, so we want to make sure we pick up from where we previously left off.

Here's a concrete scenario to think about:

Suppose we have `file_a.log`:
1.  The main agent process tells worker 1 to copy `file_a.log`.
2. Worker 1 starts copying `file_a.log`, updating the file's position in its own checkpoint file.
3. The main agent process tells worker 1 to stop copying `file_a.log` (there's multiple ways this could happen such as a configuration change removing `file_a.log` from the files that are being copied to a monitor removing it from the files that should be copied.)
4. Worker 1 stops copying `file_a.log` and removes it from its processing list.
5. The main agent process tells worker 1 (or maybe even another worker) to start copying `file_a.log` again.
6. When the main agent process goes to get the list of all checkpoints, worker 1's updates are not included in the returned values from this `get_full_checkpoints` call because `file_a.log` is no longer currently being processed.
7. The main agent process will then just use the latest checkpoint from the consolidated checkpoint file, which will not reflect the place where worker 1 had previously left off from.
8. So, when `file_a.log` starts being copied again, we could try to re-upload old bytes.

I think this is possible, correct?  We only consolidate checkpoints at specific points in time (like start up and shutdown) and I don't think all of those points cover when workers could be start and stopping the log files they are processing.

Reply:
> 3. The main agent process tells worker 1 to stop copying file_a.log (there's multiple ways this could happen such as a configuration change removing file_a.log from the files that are being copied to a monitor removing it from the files that should be copied.)

Indeed, there are multiple ways to stop a log processor.
- set a `ignore_stale_files` option for a log file so it will be closed if it will be untouched for a while.
- add and then remove the log matcher dynamically, so it also will close all its current log processors.
- 
In each of those cases, we do not mean to save the checkpoint states for the log processors that are closed.  In other words, if we remove the file and then add it again, we won't start copying from the position where it was removed but from the current position. I was trying to double-check everything and even checked this on the "pre-sharded copying manager" version and it works the same way. 

It seems that the current behavior (and behavior that was also in previous versions) may contradict the logic that you have described above, what do you think?